compiler/src/main/kotlin/dev/zacsweers/metro/compiler/ir/graph/sharding/ShardingOrchestrator.kt

```diff
@@ -0,0 +1,114 @@
+// Copyright (C) 2025 Zac Sweers
+// SPDX-License-Identifier: Apache-2.0
+package dev.zacsweers.metro.compiler.ir.graph.sharding
+
+import dev.zacsweers.metro.compiler.MetroOptions
+import dev.zacsweers.metro.compiler.graph.Component
+import dev.zacsweers.metro.compiler.graph.TopoSortResult
+import dev.zacsweers.metro.compiler.ir.IrTypeKey
+import dev.zacsweers.metro.compiler.ir.graph.DependencyGraphNode
+import dev.zacsweers.metro.compiler.isInvisibleGeneratedGraph
+
+/**
+ * Partitions bindings into shards when a graph exceeds [MetroOptions.keysPerGraphShard].
+ *
+ * Keeps SCCs together to preserve circular dependencies (including Provider<T> cycles), respects
+ * topological ordering, and merges small tail shards.
+ *
+ * Returns null if sharding isn't needed:
+ * - Graph sharding is disabled
+ * - There are too few bindings
+ * - It's either a dynamic graph or a graph extension
```

Comment:
Why is this not applicable to dynamic graphs or extensions? There will be a new type soon-ish too - anonymous graphs - that's similar to a dynamic graph but has no source declaration and are instead synthetically generated based on injection targets and input bindings/containers